CB-12384 ios: Add Cocoa Touch Framework target for CordovaLib functionality

khawkins opened the pull request:
<!--
Please make sure the checklist boxes are all checked before submitting the PR. The checklist
is intended as a quick reference, for complete details please see our Contributor Guidelines:

http://cordova.apache.org/contribute/contribute_guidelines.html

Thanks!
-->

### Platforms affected

iOS

### What does this PR do?

Adds a Cocoa Touch Framework target to the CordovaLib project, in anticipation of support for tools like Carthage (parent issue: [CB-12050](https://issues.apache.org/jira/browse/CB-12050)).

### What testing has been done on this change?

In the cordova-ios Xcode workspace under tests/, An app target was added to consume the framework target as part of the unit testing process.  All tests are verified as passing in both targets.

### Checklist
- [x] [Reported an issue](http://cordova.apache.org/contribute/issues.html) in the JIRA database
- [x] Commit message follows the format: "CB-3232: (android) Fix bug with resolving file paths", where CB-xxxx is the JIRA ID & "android" is the platform affected.
- [x] Added automated test coverage as appropriate for this change.

khawkins commented:
@shazron Is this something you could take a look at, and/or mention anyone else who might be interested?  Cc: @sgoldberg-sfdc @bhariharan

shazron commented:
This all seems sound (and seems parallel to any existing code) but I have a few clarifying questions:
1) The current CocoaPods support is for integrating CocoaPods libraries into an existing Cordova project. This PR does *not* provide the same functionality, correct? - i.e. add support for integrating Carthage libraries into an existing Cordova project.
2) I think the answer to question (1) is "Yes, it does not provide the same functionality" based on inspection of the code. So, if Cordova.framework is to be imported into an existing iOS app, how are you resolving the problem of adding Cordova plugins?

I ask about (2) since we at Adobe are trying to get an `Embedded Cordova Webview` workflow (adding Cordova to an existing iOS app) as well: https://github.com/phonegap/phonegap-webview-ios but we haven't solved the "adding Cordova plugins easily" question yet -- we just add all the core plugins.

khawkins commented:
You are correct: the support added here would only enable the consumption of the Cordova iOS dependency itself from Carthage, at this point.  Or maybe more accurately, it's the first step in enabling that support—we haven't yet verified if Carthage support will have additional requirements.

With regards to an "add Cordova plugins easily" requirement, it seems like the native side of those plugins would probably be best delivered through the dependency manager in play.  I.e. have a CocoaPods configuration for your plugin, or Carthage, or both.

Seems like the challenge would come in with the non-native pieces of the plugin, such as JS, and configuration.  This could possibly be streamlined to a degree with post-install script(s) referenced in the Podfile, with CocoaPods.  But Carthage is pretty bare bones—to my knowledge, it only supports the download and build of dependencies in a strictly Xcode-based paradigm.  Seems like that could be a lot more challenging to wrangle post-build configuration against the app that's installing the dependency.

To me, it seems like it all points to needing an amended CLI that understands/interrogates the nuances of an existing app project/workspace, whereas the current CLI is very much about orchestrating that workspace itself.  But that of course has ramifications on other platforms, etc.  Anyway, just some thoughts out loud—it's definitely a challenging problem to consider.

shazron commented:
Thanks Kevin - this is a good building block anyways, in a bigger puzzle -- we'll punt that problem off for later. I'll pull this in after I get some time to review it further.

shazron commented on tests/CordovaLibTests/CordovaLibTests.xcodeproj/project.pbxproj:
```diff
@@ -39,8 +63,29 @@
+				C0FA7CDD1E4BC5020077B045 /* Cordova.framework in Embed Frameworks */,
+			);
+			name = "Embed Frameworks";
+			runOnlyForDeploymentPostprocessing = 0;
```
This value should be 1.
This corresponds to the "Copy only when installing" checkbox in the Embed Frameworks section for the Cordova framework target in the CordovaLib project.

shazron commented on CordovaLib/CordovaLib.xcodeproj/project.pbxproj:
```diff
@@ -499,6 +637,105 @@
+				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/Frameworks @loader_path/Frameworks";
+				MTL_ENABLE_DEBUG_INFO = NO;
+				OTHER_LDFLAGS = "-all_load";
+				PRODUCT_BUNDLE_IDENTIFIER = com.apache.cordova.Cordova;
```
use `org.apache.cordova.Cordova` instead

shazron commented on CordovaLib/CordovaLib.xcodeproj/project.pbxproj:
```diff
@@ -499,6 +637,105 @@
+				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/Frameworks @loader_path/Frameworks";
+				MTL_ENABLE_DEBUG_INFO = YES;
+				OTHER_LDFLAGS = "-all_load";
+				PRODUCT_BUNDLE_IDENTIFIER = com.apache.cordova.Cordova;
```
use `org.apache.cordova.Cordova` instead

khawkins commented:
Thanks for the review, @shazron .  I've made your suggested changes, and all tests are passing through `npm test`.

shazron commented:
Thanks @khawkins  - unfortunately when I pulled other PRs in, this has caused some conflicts. Can you rebase and resolve? Sorry about this.

khawkins commented:
No worries @shazron !  Question: I see that 2a323fad2156a434e928eb570e7f272921b75d6a moved the iOS deployment target up to 9.0.  I presume I should amend the framework target to do the same?  It's currently set to 8.0 (before this change came in).

khawkins commented:
Actually, I end up seeing a ton of warnings now complaining about the fact that I didn't otherwise do that, so that probably answers my question. 😉  I'll have the update pushed shortly.

khawkins commented:
Okay @shazron , I've rebased and updated the deployment target of the framework to 9.0.  All tests are passing.  Let me know if you'd like me to squash these review commits down.

shazron commented:
Thanks @khawkins ! I'll let the CI run, then merge these in. No worries, I can squash on my end
